### PR TITLE
fix: Validate user has access to relationship type [DHIS2-17082]

### PR DESCRIPTION
**Pull request by muilpp**
When the user tries to create, update or delete a relationship, we are not validating its access to the relationship type.
This PR adds a `SecurityOwnershipValidator`, as done in TE, enrollments and events validators, to check the aforementioned access.

Ticket: https://dhis2.atlassian.net/browse/DHIS2-17082

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/RelationshipValidator.java [hunk 1]
```diff
@@ -53,7 +53,8 @@ public RelationshipValidator() {
                 new MetaValidator(),
                 new LinkValidator(),
                 new ConstraintValidator(),
-                new DuplicationValidator()));
+                new DuplicationValidator(),
+                securityOwnershipValidator));
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/RelationshipValidator.java:57]: I would put this new validator right after the `MetaValidator`.
As soon as it is validated that the relationship type exists we should check if the user has access to it.
Running this validator as the last one could potentially disclose some information that the user shouldn't see, for example, the user shouldn't be informed that there is a duplicated relationship if it has no access to the relationship type.

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java [hunk 2]
```diff
@@ -64,5 +64,6 @@ public interface RelationshipMapper extends PreheatMapper<Relationship> {
   @Mapping(target = "name")
   @Mapping(target = "attributeValues")
   @Mapping(target = "bidirectional")
+  @Mapping(target = "sharing")
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java:67]: Why are we not doing this elsewhere @enricocolasante ? And why do the other security validators work 🤷🏻 We discussed this a while back but I don't remember what the outcome was.

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  @Test
+  void shouldFailWhenUserHasNoAccessToRelationshipType() {
+    when(aclService.canDataWrite(bundle.getUser(), relationshipType)).thenReturn(false);
+
+    validator.validate(reporter, bundle, relationship);
+
+    assertNotEmpty(reporter.getErrors());
+    assertContains(
+        "has no data write access to relationship type", reporter.getErrors().get(0).getMessage());
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidatorTest.java:90]: Please use the nice assertions we have like

```java
assertHasError(reporter, enrollment, ValidationCode.E1014);
```

in org.hisp.dhis.tracker.imports.validation.validator.AssertValidations. Check other Validator tests. We should also have some now for no errors.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/bundle/RelationshipImportTest.java [hunk 4]
```diff
@@ -80,6 +82,17 @@ void successImportingRelationships() throws IOException {
     assertThat(importReport.getStats().getCreated(), is(2));
   }
 
+  @Test
+  void shouldFailWhenUserNotAuthorizedToCreateRelationship() throws IOException {
+    userA = userService.getUser("o1HMTIzBGo7");
+    TrackerImportParams params = TrackerImportParams.builder().userId(userA.getUid()).build();
+    ImportReport importReport =
+        trackerImportService.importTracker(params, fromJson("tracker/relationships.json"));
+    assertThat(importReport.getStatus(), is(Status.ERROR));
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/bundle/RelationshipImportTest.java:91]: that is already asserted by assertHasError

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java (hunk 2 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java:67]: I am not sure I understand what you mean by elsewhere.
We are mapping `sharing` for other metadata, like `TrackedEntityType` or `Program`

Comment (muilpp) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java]: I think @teleivo meant why are we not using the `RelationshipMapperType` like we do in the `TrackerMapper` for the TET.
I refactored this in the last commit: https://github.com/dhis2/dhis2-core/pull/16880/commits/d921f3d4b014bde82d42d636525046b8fdfdd77c

File: dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidatorTest.java [hunk 5]
```diff
@@ -70,25 +73,25 @@ public void setUp() {
...
     reporter = new Reporter(idSchemes);
     relationship = new Relationship();
+    relationshipType = new RelationshipType();
+
+    when(bundle.getPreheat().getRelationshipType(any())).thenReturn(relationshipType);
+    when(bundle.getStrategy(relationship)).thenReturn(CREATE);
   }
 
   @Test
   void shouldFailWhenUserHasNoAccessToRelationshipType() {
```
Comment (lucaCambi77) [inline dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidatorTest.java]: Are we covering the different test cases? I understand this validation happens when the strategy is update/delete/create. Also, I see the integration and JSON cover only the negative case, but I don't have a full vision of what is already present

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidator.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import org.hisp.dhis.tracker.imports.domain.Relationship;
+import org.hisp.dhis.tracker.imports.validation.Reporter;
+import org.hisp.dhis.tracker.imports.validation.ValidationCode;
+import org.hisp.dhis.tracker.imports.validation.Validator;
+import org.springframework.stereotype.Component;
+
+@Component(
+    "org.hisp.dhis.tracker.imports.validation.validator.relationship.SecurityOwnershipValidator")
+@RequiredArgsConstructor
+@Slf4j
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/relationship/SecurityOwnershipValidator.java:46]: No need to add the logging dependency if you don't log.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/bundle/RelationshipImportTest.java (hunk 4 above)
Comment (teleivo) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/bundle/RelationshipImportTest.java:91]: `assertThat(importReport.getStatus(), is(Status.ERROR));` is already asserted by `assertHasError`. Can you please remove that line.

And add newline between setup, method under test and assertions. So it's easy to spot them.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java (hunk 2 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/preheat/mappers/RelationshipMapper.java]: Perfect! Now it's clear that we map the sharing for metadata. Before I thought we map it now for data which we do not do elsewhere.

**Review by teleivo — APPROVED** [review]

**Pull request merged**